classes/Datasets/Transforms.sc

```diff
@@ -0,0 +1,122 @@
+/*****************************************
+Data Transforms
+(C) 2018 Jonathan Reus
+
+Tools for analyzing and transforming datasets: scaling, normalization, standardization, PCA, etc..
+
+
+******************************************/
+
+// TO NORMALIZE:
+// find min / max of features
+// normalized_value = (val - min) / (max - min)
+
+
+/*
+T for Transform
+Normalization:
+1. find min/max of each feature
+2. normalized_value = (val - min) / (max - min)
+*/
+TNormalizer {
+	var <min, <max, <originalData, <normalizedData, dim;
+
+	/*
+	@param dataset a Matrix of rows as feature vectors
+	*/
+	*new {|dataset|
+		^super.new.init(dataset);
+	}
+
+	init {|dataset|
+		if(dataset.isKindOf(SequenceableCollection).not ) { "Dataset must be a Matrix".error;this.halt; };
+		if(dataset.at(0).isKindOf(SequenceableCollection)) {
+			dim = dataset.cols;
+			min = Array.newClear(dim);
+			max = Array.newClear(dim);
+			normalizedData = Matrix.newClear(dataset.rows, dataset.cols);
+			dim.do {|i|
+				var col = dataset.getCol(i);
+				min[i] = col.minItem;
+				max[i] = col.maxItem;
+				normalizedData.putCol(i, this.normalizeSample(col) );
+			};
+		} {
+			min = dataset.minItem; max = dataset.maxItem;
+			normalizedData = dataset.collect {|it,i| this.normalizeSample(it) };
+		};
+		originalData = dataset;
+	}
+
+	// normalize / denormalize a single sample
+	normalizeSample {|samp| (samp - min) / (max - min) }
+	denormalizeSample {|samp| (samp * (max-min)) + min }
+
+
+	// normalize an entire dataset
+	normalizeData {}
+
+	// denormalize a point-slope form 2-dimensional line
+	// of the form [slope, intercept]
+	denormalizeLine {|line|
```

Comment:
can you check the formatting of whitespace? Please use tabs not spaces, and auto-formatting.